node.go

```
@@ -64,69 +66,39 @@ func NewInteriorNode(parent MerkleNode, level int, prefixBits []bool) *interiorN
 	return newNode
 }
 
-type MerkleNode interface {
-	Value() []byte
-	isEmpty() bool
-}
-
-var _ MerkleNode = (*userLeafNode)(nil)
-var _ MerkleNode = (*interiorNode)(nil)
-var _ MerkleNode = (*emptyNode)(nil)
-
-func (n *userLeafNode) Value() []byte {
-	return n.value
-}
-
-func (n *userLeafNode) isEmpty() bool {
-	return false
-}
-
-func (n *interiorNode) Value() []byte {
-	return nil
-}
-
-func (n *interiorNode) isEmpty() bool {
-	return false
-}
-
-func (n *emptyNode) Value() []byte {
-	return nil
-}
-
-func (n *emptyNode) isEmpty() bool {
-	return true
-}
-
-type ProofNode interface {
-	Commitment() []byte
-}
-
-var _ ProofNode = (*userLeafNode)(nil)
-var _ ProofNode = (*emptyNode)(nil)
-
-func (n *emptyNode) Commitment() []byte {
-	return nil
-}
-
-func (n *userLeafNode) Commitment() []byte {
-	return n.commitment
-}
-
 func (n *interiorNode) serialize() []byte {
 	input := make([]byte, 0, len(n.leftHash)+len(n.rightHash))
 	input = append(input, n.leftHash...)
 	input = append(input, n.rightHash...)
 	return input
 }
 
-func (n *interiorNode) hash() []byte {
+func hashNode(node MerkleNode, m *MerkleTree) []byte {
+	switch node.(type) {
+	case *interiorNode:
+		return node.(*interiorNode).hash(m)
+	case *userLeafNode:
+		return node.(*userLeafNode).hash(m)
+	case *emptyNode:
+		return node.(*emptyNode).hash(m)
+	}
+	return nil
```

panic, instead of returning nil